src/profile-logic/transforms.js

```diff
@@ -1235,85 +1235,33 @@ const FUNC_MATCHES = {
 
 export function collapseFunctionSubtree(
   thread: Thread,
-  funcToCollapse: IndexIntoFuncTable,
-  defaultCategory: IndexIntoCategoryList
+  funcToCollapse: IndexIntoFuncTable
 ): Thread {
   const { stackTable, frameTable } = thread;
-  const oldStackToNewStack: Map<
-    IndexIntoStackTable | null,
-    IndexIntoStackTable | null,
-  > = new Map();
-  // A root stack's prefix will be null. Maintain that relationship from old to new
-  // stacks by mapping from null to null.
-  oldStackToNewStack.set(null, null);
-  const collapsedStacks = new Set();
-  const newStackTable = getEmptyStackTable();
+  const oldStackToNewStack = new Int32Array(stackTable.length);
+  const isInCollapsedSubtree = new Uint8Array(stackTable.length);
 
   for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
     const prefix = stackTable.prefix[stackIndex];
-    if (
-      // The previous stack was collapsed, this one is collapsed too.
-      collapsedStacks.has(prefix)
-    ) {
-      // Only remember that this stack is collapsed.
-      const newPrefixStackIndex = oldStackToNewStack.get(prefix);
-      if (newPrefixStackIndex === undefined) {
-        throw new Error('newPrefixStackIndex cannot be undefined');
-      }
-      // Many collapsed stacks will potentially all point to the first stack that used the
-      // funcToCollapse, so newPrefixStackIndex will potentially be assigned to many
-      // stacks. This is what actually "collapses" a stack.
-      oldStackToNewStack.set(stackIndex, newPrefixStackIndex);
-      collapsedStacks.add(stackIndex);
-
-      // Fall back to the default category when stack categories conflict.
-      if (newPrefixStackIndex !== null) {
-        if (
-          newStackTable.category[newPrefixStackIndex] !==
-          stackTable.category[stackIndex]
-        ) {
-          newStackTable.category[newPrefixStackIndex] = defaultCategory;
-          newStackTable.subcategory[newPrefixStackIndex] = 0;
-        } else if (
-          newStackTable.subcategory[newPrefixStackIndex] !==
-          stackTable.subcategory[stackIndex]
-        ) {
-          newStackTable.subcategory[newPrefixStackIndex] = 0;
-        }
-      }
+    if (prefix !== null && isInCollapsedSubtree[prefix] !== 0) {
```

Comment:
The background is:

 - The call node table has to use -1, because it uses an Int32Array, so it can't store null members.
 - The stack table is a regular array, so it can store `null`, and `null` allows the type system to catch more mistakes.
 - The stack table is not a typed array because its size is often not known upfront, and because we'd need to do more work to serialize it to / from JSON.

An Int32Array has the advantage over a regular array that it takes only half the amount of memory on 64-bit machines, i.e. 4 bytes per element instead of 8 bytes. Most algorithms which operate on the call node table are memory bandwidth bound, so halving the size makes the algorithms run in half the time.

Recently I've learned that, for regular arrays, WebKit and Chrome (actually JSC and V8) have optimizations for arrays which only contain numbers. So even for regular arrays it may make sense to use -1 instead of `null`. It is my understanding that Chrome heap-allocates numbers unless the numbers are stored in a number-only array, so making this change might save memory and improve performance in Chrome. And [this WebKit code](https://searchfox.org/wubkat/rev/fdf03b86449c12a28214ee47eee06577e538f4e3/Source/JavaScriptCore/runtime/IndexingType.h#107-109) suggests that they can special-case both array-of-int32 and array-of-double. These special cases may also have positive effects on GC performance: If you know that the contents of a big array cannot point to other objects, then you do not have to traverse that memory during the mark phase.